src/fetchers.ts

```diff
@@ -282,6 +297,48 @@ async function fetchHelm(
   });
 }
 
+async function fetchNixpkgs(
+  spec: NixpkgsFetch,
+  context: FetchContext,
+): Promise<Version[]> {
+  const nixexprs = await fetchUrl("https://channels.nixos.org/" + spec.channel, context, undefined, { redirect: "manual" }, async (response) => {
+    if (response.status !== 301) {
+      throw new Error(
+        `Got status ${response.status}: ${await response.text()}`,
+      );
+    }
+    const location = response.headers.get("Location");
+    if (location === null) {
+      throw new Error(`Could not get location for ${response.url}`);
+    }
+    return location + "/nixexprs.tar.xz";
+  });
+  let hash = undefined as string | undefined;
+  if (!context.update) {
+    const cached = context.cache.getBody(nixexprs);
+    if (cached) {
+      hash = cached;
+    }
+  }
+  if (hash === undefined) {
+    const p = Deno.run({
+      cmd: [
+        "nix-prefetch-url",
+        "--unpack",
+        nixexprs,
+      ],
+      stdout: "piped",
+    });
+    const [{ code }, output] = await Promise.all([p.status(), p.output()]);
+    if (code !== 0) {
+      throw new Error(`Failed to fetch hash for ${nixexprs}`);
+    }
+    hash = new TextDecoder().decode(output).trim();
+    context.cache.saveBody(nixexprs, hash);
+  }
+  return [makeVersion(hash)];
```

Comment:
This fetcher will only return one kind of version, so you could skip the inference and return a `NixHashVersion` (or a hypothetical `NixChannelVersion` directly).

This probably applies to the rest of the file too: the `fetchGithubCommit` function is only going to return a `CommitVersion`, so passing it through `makeVersion` requires `makeVersion` to infer if it's a commit or not. You don't have to fix the rest of the file though.